enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,204 @@
+  - Configmaps annotated for CA bundle injection should be updated with the new CA bundle.
+  - API services annotated for CA bundle injection should be updated with the new CA bundle.
+
+### Stretch Goals
+
+- Support forced rotation
+  - Rotation triggered by an API change rather than the validity bounds of the CA
+
+- Support forced rotation with a custom validity bounds to support emergency rotation
+  - i.e. support minimizing the grace period for trusting the pre-rotation CA
+
+### Non-Goals
+
+- Support forced rotation using an external CA and signing key. This could be
+  useful for admins to plug in their own intermediate CA and have all serving
+  certs issued by it, and also for future integration efforts.
+
+- Implement library-go support for [hitless
+rotation](https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/)
+so that consumers of serving certs and CA bundles can trivially support automated
+refresh.
+
+## Proposal
+
+- Detect when the current CA certificate has passed 1/2 of its validity bounds
+  (i.e. 6 months has passed of a 1 year validity) and generate a new CA.
+  - The new CA should have the same `Subject.CommonName` as the current CA to
+    ensure that trust chaining is possible.
+  - This provides a 6 month grace period for consumers of serving certs and ca
+    bundles to be refreshed.
+
+- Include both the new and old CAs' primary certificate in a new CA bundle.
+  - The previous CA's primary certificate will serve to bridge trust between an
+    unrefreshed server (serving with a cert generated from the previous CA) and
+    a refreshed client (validating trust with the new CA bundle). The existing
+    CA's primary certificate will ensure that certificates generated from the
+    new CA can be trusted.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
+  but signed with the private key of the current CA.
+  - This intermediate certificate should be included with newly-generated serving
+    certificates, and serves to bridge trust between the CA bundle of an
+    unrefreshed client (validating trust with the old CA bundle) and a service
+    serving with a cert generated from the new CA.
+
+- Compare the `AuthorityKeyId` of a serving cert with the `SubjectKeyId` of the
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+
+- Provide an indication that rotation has taken place in the status of the
```

Comment:
would expect events not a condition. Conditions are level, events are edges.

Reply:
Done